### PR TITLE
feat: add trainer udf

### PR DESCRIPTION
**Pull request by ab93**

### CODE REVIEW

File: numalogic/tools/types.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
 
 artifact_t = TypeVar(
     "artifact_t",
-    bound=Union[TorchModel, BaseThresholdModel, BaseTransformer],
```
Comment (s0nicboOm) [inline numalogic/tools/types.py:30]: Is this addition for preproc_clf?

File: numalogic/udfs/trainer.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+            model,
+            x_train,
+            preproc_clf=preproc_clf,
+            threshold_clf=thresh_clf,
+            trainer_cfg=_conf.numalogic_conf.trainer,
+        )
+
+        # Save artifacts
+        # TODO perform multi-save here
+        self.save_artifact(
```
Comment (s0nicboOm) [inline numalogic/udfs/trainer.py:187]: Need to save preproc only if stateful is true for one of the clf's

File: numalogic/udfs/trainer.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+        Args:
+            artifact: Artifact to save
+            skeys: List of keys
+            dkeys: List of dkeys
+            uuid: UUID
+            **metadata: Additional metadata
+        """
+        if not artifact:
+            return
+        if isinstance(artifact, StatelessTransformer):
```
Comment (s0nicboOm) [inline numalogic/udfs/trainer.py:240]: For preproc maybe we still might to add checks cause it will return a sklearn pipeline object?

**Review by s0nicboOm — APPROVED** [review]
Minor comments

File: numalogic/tools/types.py (hunk 1 above)
Comment (ab93) [reply numalogic/tools/types.py:30]: Correct

**Pull request merged**